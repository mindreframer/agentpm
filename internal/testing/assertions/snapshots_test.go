package assertions

import (
	"fmt"
	"testing"

	"github.com/mindreframer/agentpm/internal/testing/builders"
	"github.com/mindreframer/agentpm/internal/testing/executor"
)

// Phase 4B: Snapshot Integration Tests

func TestSnapshotIntegration_CapturesFullStateCorrectly(t *testing.T) {
	// Create environment and epic for snapshot testing
	env := executor.NewTestExecutionEnvironment("snapshot-test.xml")

	testEpic, err := builders.NewEpicBuilder("snapshot-test").
		WithStatus("planning").
		WithPhase("1A", "Setup", "pending").
		WithTask("1A_1", "1A", "Initialize", "pending").
		WithTest("T1A_1", "1A_1", "1A", "Test Init", "pending").
		Build()

	if err != nil {
		t.Fatalf("Failed to build test epic: %v", err)
	}

	err = env.LoadEpic(testEpic)
	if err != nil {
		t.Fatalf("Failed to load epic: %v", err)
	}

	// Execute basic workflow
	result, err := executor.CreateTransitionChain(env).
		StartEpic().
		StartPhase("1A").
		StartTask("1A_1").
		PassTest("T1A_1").
		DoneTask("1A_1").
		DonePhase("1A").
		Execute()

	if err != nil {
		t.Fatalf("Failed to execute transition chain: %v", err)
	}

	// Test snapshot integration captures full state with timestamp normalization
	snapshotAssertion := NewSnapshotAssertionWithConfig(t, true, false) // Enable cross-platform, disable update

	err = snapshotAssertion.MatchSnapshot("full_state_test", result.FinalState)
	if err != nil {
		t.Errorf("Expected snapshot to capture full state correctly, got error: %v", err)
	}
}

func TestSnapshotIntegration_XMLSnapshotsShowMeaningfulDiffs(t *testing.T) {
	// Test XML snapshots with different content
	xmlContent1 := `<epic status="active"><phase id="1A" status="completed"/></epic>`
	xmlContent2 := `<epic status="pending"><phase id="1A" status="active"/></epic>`

	snapshotAssertion := NewSnapshotAssertion(t)

	// First snapshot should succeed
	err := snapshotAssertion.MatchXMLSnapshot("xml_diff_test", xmlContent1)
	if err != nil {
		t.Errorf("First XML snapshot failed: %v", err)
	}

	// Test that different XML shows meaningful diff
	// (In real usage, the second call would show a diff)
	err = snapshotAssertion.MatchXMLSnapshot("xml_diff_test_2", xmlContent2)
	if err != nil {
		t.Errorf("Second XML snapshot failed: %v", err)
	}
}

func TestSnapshotIntegration_SnapshotUpdatesWorkDuringDevelopment(t *testing.T) {
	// Test snapshot update mechanism
	snapshotAssertion := NewSnapshotAssertionWithConfig(t, false, true) // Enable update mode

	testData := map[string]interface{}{
		"epic_id":     "update-test",
		"epic_status": "active",
		"phases":      2,
		"tasks":       4,
		"tests":       4,
	}

	err := snapshotAssertion.MatchSnapshot("update_test", testData)
	if err != nil {
		t.Errorf("Expected snapshot update to work, got error: %v", err)
	}

	// Test that update mode affects behavior
	normalizer := snapshotAssertion.normalizer.(*DefaultSnapshotNormalizer)
	if !normalizer.updateMode {
		t.Error("Expected update mode to be enabled")
	}
}

func TestSnapshotIntegration_SelectiveSnapshotsFocusOnRelevantElements(t *testing.T) {
	// Create complex test data
	testData := map[string]interface{}{
		"epic_id":        "selective-test",
		"epic_status":    "active",
		"phases":         3,
		"tasks":          6,
		"tests":          8,
		"events":         15,
		"execution_time": "125ms",
		"memory_usage":   "45MB",
		"internal_data":  "should_be_filtered",
	}

	snapshotAssertion := NewSnapshotAssertion(t)

	// Test selective snapshot focusing on relevant fields
	selectedFields := []string{"epic_id", "epic_status", "phases", "tasks", "tests"}

	err := snapshotAssertion.MatchSelectiveSnapshot("selective_test", testData, selectedFields)
	if err != nil {
		t.Errorf("Expected selective snapshot to work, got error: %v", err)
	}
}

func TestSnapshotIntegration_CrossPlatformSnapshotsAreConsistent(t *testing.T) {
	// Test cross-platform normalization
	testDataWithPaths := map[string]interface{}{
		"epic_id":     "cross-platform-test",
		"file_path":   "/Users/test/file.xml",   // Unix-style path
		"work_dir":    "C:\\Projects\\AgentPM",  // Windows-style path
		"config":      "./config/settings.json", // Relative path
		"epic_status": "active",
	}

	snapshotAssertion := NewSnapshotAssertionWithConfig(t, true, false) // Enable cross-platform

	err := snapshotAssertion.MatchSnapshot("cross_platform_test", testDataWithPaths)
	if err != nil {
		t.Errorf("Expected cross-platform snapshot to work, got error: %v", err)
	}

	// Test that cross-platform normalization is applied
	normalizer := snapshotAssertion.normalizer.(*DefaultSnapshotNormalizer)
	if !normalizer.crossPlatform {
		t.Error("Expected cross-platform mode to be enabled")
	}

	// Test normalization behavior
	normalized := normalizer.applyCrossPlatformNormalization(testDataWithPaths)
	if normalizedData, ok := normalized.(map[string]interface{}); ok {
		// Verify that normalization was applied (paths should be consistent)
		if normalizedData["epic_id"] != "cross-platform-test" {
			t.Error("Cross-platform normalization affected non-path data")
		}
	}
}

func TestSnapshotIntegration_ConfigurationOptions(t *testing.T) {
	// Test various configuration combinations
	testConfigs := []struct {
		name          string
		crossPlatform bool
		updateMode    bool
	}{
		{"default", false, false},
		{"cross_platform_only", true, false},
		{"update_mode_only", false, true},
		{"both_enabled", true, true},
	}

	testData := map[string]interface{}{
		"test_id": "config-test",
		"status":  "active",
	}

	for _, config := range testConfigs {
		t.Run(config.name, func(t *testing.T) {
			snapshotAssertion := NewSnapshotAssertionWithConfig(t, config.crossPlatform, config.updateMode)

			err := snapshotAssertion.MatchSnapshot(config.name, testData)
			if err != nil {
				t.Errorf("Configuration %s failed: %v", config.name, err)
			}

			// Verify configuration was applied
			normalizer := snapshotAssertion.normalizer.(*DefaultSnapshotNormalizer)
			if normalizer.crossPlatform != config.crossPlatform {
				t.Errorf("Expected crossPlatform=%v, got %v", config.crossPlatform, normalizer.crossPlatform)
			}
			if normalizer.updateMode != config.updateMode {
				t.Errorf("Expected updateMode=%v, got %v", config.updateMode, normalizer.updateMode)
			}
		})
	}
}

func TestSnapshotIntegration_XMLNormalization(t *testing.T) {
	// Test XML normalization for consistent snapshots
	xmlVariations := []string{
		`<epic status="active"><phase id="1A"/></epic>`,
		`<epic   status="active"  ><phase   id="1A"   /></epic>`, // Extra whitespace
		`<epic status='active'><phase id='1A'/></epic>`,          // Single quotes
		`<epic status="active">
			<phase id="1A"/>
		</epic>`, // Multi-line formatting
	}

	snapshotAssertion := NewSnapshotAssertion(t)

	for i, xml := range xmlVariations {
		testName := fmt.Sprintf("xml_norm_%d", i)
		err := snapshotAssertion.MatchXMLSnapshot(testName, xml)
		if err != nil {
			t.Errorf("XML normalization test %d failed: %v", i, err)
		}
	}
}

func TestSnapshotIntegration_ErrorHandling(t *testing.T) {
	// Test error handling in snapshot operations
	snapshotAssertion := NewSnapshotAssertion(nil) // No testing.T

	err := snapshotAssertion.MatchSnapshot("error_test", "test data")
	if err == nil {
		t.Error("Expected error when testing.T is nil")
	}

	if !containsString(err.Error(), "no testing.T instance") {
		t.Errorf("Expected specific error message, got: %v", err)
	}
}

// Helper function to check if string contains substring (using simple approach)
func containsString(s, substr string) bool {
	for i := 0; i <= len(s)-len(substr); i++ {
		if s[i:i+len(substr)] == substr {
			return true
		}
	}
	return false
}
