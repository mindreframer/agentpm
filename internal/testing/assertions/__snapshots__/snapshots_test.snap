
[TestSnapshotIntegration_CapturesFullStateCorrectly - 1]
full_state_test
map[string]interface {}{
    "Assignee":     "",
    "CreatedAt":    "NORMALIZED_TIMESTAMP",
    "CurrentState": map[string]interface {}{
        "ActivePhase": "",
        "ActiveTask":  "",
        "NextAction":  "Start next phase",
    },
    "Dependencies": "",
    "Description":  "",
    "Events":       []interface {}{
        map[string]interface {}{
            "Data":      "Epic snapshot-test started",
            "ID":        "epic_started_NORMALIZED_TIMESTAMP",
            "Timestamp": "NORMALIZED_TIMESTAMP",
            "Type":      "epic_started",
        },
        map[string]interface {}{
            "Data":      "Phase 1A (Setup) started",
            "ID":        "phase_started_NORMALIZED_TIMESTAMP",
            "Timestamp": "NORMALIZED_TIMESTAMP",
            "Type":      "phase_started",
        },
        map[string]interface {}{
            "Data":      "Task 1A_1 (Initialize) started",
            "ID":        "task_started_NORMALIZED_TIMESTAMP",
            "Timestamp": "NORMALIZED_TIMESTAMP",
            "Type":      "task_started",
        },
        map[string]interface {}{
            "Data":      "Test T1A_1 passed",
            "ID":        "test_passed_T1A_1_NORMALIZED_TIMESTAMP",
            "Timestamp": "NORMALIZED_TIMESTAMP",
            "Type":      "test_passed",
        },
        map[string]interface {}{
            "Data":      "Task 1A_1 (Initialize) completed",
            "ID":        "task_completed_NORMALIZED_TIMESTAMP",
            "Timestamp": "NORMALIZED_TIMESTAMP",
            "Type":      "task_completed",
        },
        map[string]interface {}{
            "Data":      "Phase 1A (Setup) completed",
            "ID":        "phase_completed_NORMALIZED_TIMESTAMP",
            "Timestamp": "NORMALIZED_TIMESTAMP",
            "Type":      "phase_completed",
        },
    },
    "ID":       "snapshot-test",
    "Metadata": map[string]interface {}{
        "Assignee":        "",
        "Created":         "NORMALIZED_TIMESTAMP",
        "EstimatedEffort": "",
    },
    "Name":   "snapshot-test",
    "Phases": []interface {}{
        map[string]interface {}{
            "CompletedAt":  "NORMALIZED_TIMESTAMP",
            "Deliverables": "",
            "Description":  "",
            "ID":           "1A",
            "Name":         "Setup",
            "StartedAt":    "NORMALIZED_TIMESTAMP",
            "Status":       "completed",
        },
    },
    "Requirements": "",
    "Status":       "wip",
    "Tasks":        []interface {}{
        map[string]interface {}{
            "AcceptanceCriteria": "",
            "Assignee":           "",
            "CancelledAt":        nil,
            "CompletedAt":        "NORMALIZED_TIMESTAMP",
            "Description":        "",
            "ID":                 "1A_1",
            "Name":               "Initialize",
            "PhaseID":            "1A",
            "StartedAt":          "NORMALIZED_TIMESTAMP",
            "Status":             "completed",
        },
    },
    "Tests": []interface {}{
        map[string]interface {}{
            "CancellationReason": "",
            "CancelledAt":        nil,
            "Description":        "",
            "FailedAt":           nil,
            "FailureNote":        "",
            "ID":                 "T1A_1",
            "Name":               "Test Init",
            "PassedAt":           "NORMALIZED_TIMESTAMP",
            "PhaseID":            "1A",
            "StartedAt":          nil,
            "Status":             "completed",
            "TaskID":             "1A_1",
            "TestResult":         "passing",
            "TestStatus":         "done",
        },
    },
    "Workflow": "",
}
---

[TestSnapshotIntegration_XMLSnapshotsShowMeaningfulDiffs - 1]
xml_diff_test
<epic status="wip"><phase id="1A" status="completed"/></epic>
---

[TestSnapshotIntegration_XMLSnapshotsShowMeaningfulDiffs - 2]
xml_diff_test_2
<epic status="pending"><phase id="1A" status="wip"/></epic>
---

[TestSnapshotIntegration_SnapshotUpdatesWorkDuringDevelopment - 1]
update_test
map[string]interface {}{
    "epic_id":     "update-test",
    "epic_status": "wip",
    "phases":      float64(2),
    "tasks":       float64(4),
    "tests":       float64(4),
}
---

[TestSnapshotIntegration_SelectiveSnapshotsFocusOnRelevantElements - 1]
selective_test
map[string]interface {}{
    "epic_id":     "selective-test",
    "epic_status": "wip",
    "phases":      float64(3),
    "tasks":       float64(6),
    "tests":       float64(8),
}
---

[TestSnapshotIntegration_CrossPlatformSnapshotsAreConsistent - 1]
cross_platform_test
map[string]interface {}{
    "config":      "./config/settings.json",
    "epic_id":     "cross-platform-test",
    "epic_status": "wip",
    "file_path":   "/Users/test/file.xml",
    "work_dir":    "C:\\Projects\\AgentPM",
}
---

[TestSnapshotIntegration_ConfigurationOptions/default - 1]
default
map[string]interface {}{
    "status":  "wip",
    "test_id": "config-test",
}
---

[TestSnapshotIntegration_ConfigurationOptions/cross_platform_only - 1]
cross_platform_only
map[string]interface {}{
    "status":  "wip",
    "test_id": "config-test",
}
---

[TestSnapshotIntegration_ConfigurationOptions/update_mode_only - 1]
update_mode_only
map[string]interface {}{
    "status":  "wip",
    "test_id": "config-test",
}
---

[TestSnapshotIntegration_ConfigurationOptions/both_enabled - 1]
both_enabled
map[string]interface {}{
    "status":  "wip",
    "test_id": "config-test",
}
---

[TestSnapshotIntegration_XMLNormalization - 1]
xml_norm_0
<epic status="wip"><phase id="1A"/></epic>
---

[TestSnapshotIntegration_XMLNormalization - 2]
xml_norm_1
<epic   status="wip"  ><phase   id="1A"   /></epic>
---

[TestSnapshotIntegration_XMLNormalization - 3]
xml_norm_2
<epic status='wip'><phase id='1A'/></epic>
---

[TestSnapshotIntegration_XMLNormalization - 4]
xml_norm_3
<epic status="wip">
            <phase id="1A"/>
        </epic>
---
